### DISCUSSION TITLE
api.tree.toggle doesn't work when setup is called right before it in a mapping

### DISCUSSION THREAD

**Discussion by bernardassan** [nvim-tree/nvim-tree.lua#2052] [Q&A] [answered]
### Description

I am trying to manually lazy load nvim-tree on a mapping `<leader>n`, so I setup config like normally would but moved setup to be called in the mapping function. But I realized that when setup is called, followed by `api.tree.toggle` in the map, it always opens the tree window and never closes it. __NOTE: but if I call the setup function outside of the mapping function .ie `vim.keymap.set` everything works normal__ . But because I am trying to lazy load nvim-tree, calling setup outside of the map isn't an option. In addition, the `api.tree.open` and `api.tree.close` work perfectly when mapped (.ie setup function then api.tree.*) only toggle isn't working for me.

I think this issue is manifest in `view.is_visible()` in the `toggle fn`

### Neovim version

```text
NVIM v0.8.3
Build type: Release
LuaJIT 2.1.0-beta3
Compilado por builduser

Features: +acl +iconv +tui
See ":help feature-compile"

     archivo "vimrc" del sistema: "$VIM/sysinit.vim"
            predefinido para $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

### Operating system and version

Archlinux, Linux malpha 6.2.2-zen2-1-zen #1 ZEN SMP PREEMPT_DYNAMIC Wed, 08 Mar 2023 04:07:25 +0000 x86_64 GNU/Linux

### nvim-tree version

bbb6d48

### Minimal config

```lua
vim.cmd [[set runtimepath=$VIMRUNTIME]]
vim.cmd [[set packpath=/tmp/nvt-min/site]]
local package_root = "/tmp/nvt-min/site/pack"
local install_path = package_root .. "/packer/start/packer.nvim"
local function load_plugins()
  require("packer").startup {
    {
      "wbthomason/packer.nvim",
      "nvim-tree/nvim-tree.lua",
      "nvim-tree/nvim-web-devicons",
      -- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
    },
    config = {
      package_root = package_root,
      compile_path = install_path .. "/plugin/packer_compiled.lua",
      display = { non_interactive = true },
    },
  }
end
if vim.fn.isdirectory(install_path) == 0 then
  print "Installing nvim-tree and dependencies."
  vim.fn.system { "git", "clone", "--depth=1", "https://github.com/wbthomason/packer.nvim", install_path }
end
load_plugins()
require("packer").sync()
vim.cmd [[autocmd User PackerComplete ++once echo "Ready!" | lua setup()]]
vim.opt.termguicolors = true
vim.opt.cursorline = true

-- MODIFY NVIM-TREE SETTINGS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
_G.setup = function()
 vim.keymap.set("n", ",n", function()
    require("nvim-tree").setup()
    return require("nvim-tree.api").tree.toggle({ focus = false, find_file = true })
end
, { silent = true })

end
```

### Steps to reproduce

1. nvim -nu /tmp/nvt-min.lua
2. while in normal mode in editor, press `,n`
3. NvimTree opens
4. press `,n` again to toggle the tree. The NvimTree doesn't close but refreshes/opens again.

### Expected behavior

I expected that the `,n` mapping toggles the window .ie it opens it first and closes it afterward.

### Actual behavior

The tree opens as expected but the next `,n` press doesn't close/toggle the tree

**Comment by alex-courtis** [comment]
nvim-tree setup is intended to be called once at startup and again only if you have made configuration changes.

The first call is cheap, doing nothing more than validating and applying the configuration. Nothing happens until the tree is opened.
The subsequent calls are quite expensive, as they must tear down the world first.

Short answer: don't do this. Call setup once in your `init.lua` rather than your mapping.

Lazy loading: if you have a use case that absolutely requires lazy loading, please share it. I've not received a good answer other than "the cool factor".

Long answer: see `:help nvim_tree_events_startup` for two mechanisms you can use to determine whether nvim-tree has been setup, to conditionally call setup in your mapping.

**Reply by alex-courtis** [reply]
Clarified documentation for nvim-tree-setup:

https://github.com/nvim-tree/nvim-tree.lua/blob/master/doc/nvim-tree-lua.txt#L170

**Comment by nitaicharan** [comment]
It would be an interest feature, @alex-courtis . I'm also having this problem after migrate to [lazy.nvim](https://github.com/folke/lazy.nvim) package manager. The idea is, only load the package (in this case the nvim-tree) if you need it. It allows a faster neovim bootstrap

Following my configuration:

```lua
local tree_interface = require("interfaces.tree")

local _setup = function()
	vim.g.loaded_netrw = 1
	vim.g.loaded_netrwPlugin = 1
	vim.opt.termguicolors = true

	require("nvim-tree").setup({
		sync_root_with_cwd = true,
		respect_buf_cwd = true,
		view = {
			mappings = {
				custom_only = true,
				list = {
					{ key = "d", action = "remove" },
					{ key = "h", action = "parent_node" },
					{ key = "H", action = "close_node" },
					{ key = "R", action = "rename" },
					{ key = "m", action = "bulk_move" },
					{ key = "p", action = "paste" },
					{ key = "P", action = "preview" },
					{ key = "cf", action = "create" },
					{ key = "gr", action = "refresh" },
					{ key = "oh", action = "split" },
					{ key = "ov", action = "vsplit" },
					{ key = "ox", action = "system_open" },
					{ key = "yf", action = "copy" },
					{ key = "ya", action = "copy_absolute_path" },
					{ key = "yp", action = "copy_name" },
					{ key = "yr", action = "copy_path" },
					{ key = { "oo", "l" }, action = "edit" },
					{ key = "?", action = "toggle_help" },
				},
			},
		},
		renderer = {
			highlight_git = true,
			root_folder_label = ":t",
			icons = {
				show = {
					git = false,
				},
			},
		},
		update_focused_file = {
			enable = true,
			update_root = true,
			debounce_delay = 15,
		},
		diagnostics = {
			enable = true,
			show_on_open_dirs = true,
		},
		git = {
			ignore = false,
			timeout = 200,
		},
		trash = {
			cmd = "trash",
		},
	})
end

tree_interface.open = function()
	_setup()

	print("open")
	require("nvim-tree.api").tree.open()
end

tree_interface.close = function()
	_setup()

	print("close")
	require("nvim-tree.api").tree.close()
end

tree_interface.toggle = function()
	_setup()

	print("toggle")
	require("nvim-tree.api").tree.toggle()
end

return tree_interface
```

**Reply by nitaicharan** [reply]
P.S. The `print("???")` I placed to see if I was actually calling the right function on my [which-key.nvim](https://github.com/folke/which-key.nvim) config mapping after many times trying to figure out the mistake.

**Reply by alex-courtis** [reply]
>  The idea is, only load the package (in this case the nvim-tree) if you need it. It allows a faster neovim bootstrap

I do not think this is correct. Lua packages are only loaded when they are `require`d, not when they are added to `&packpath` / `&runtimepath` 

Have you benchmarked this? I am happy to be proven wrong.

> vim.g.loaded_netrw = 1
> vim.g.loaded_netrwPlugin = 1

Do this at the very start of your init.lua. By the time you hit your function it will be too late and netrw will have loaded. It's not a regular plugin and starts uncontrollably eagerly.

> I'm also having this problem after migrate to [lazy.nvim](https://github.com/folke/lazy.nvim) package manager.

Testing with the [clean room configuration](https://github.com/nvim-tree/nvim-tree.lua/wiki/Troubleshooting#clean-room-replication) calling setup then open was successful.

```lua
-- MODIFY NVIM-TREE SETTINGS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
_G.setup = function()
  require("nvim-tree").setup {}
  require("nvim-tree.api").tree.open()
end
```

It looks like you are calling setup many times. Only call it when applying the initial configuration or to apply a new configuration. Subsequent calls tear down the world.

**Reply by nitaicharan** [reply]
Thank you to replay it.

> I do not think this is correct. Lua packages are only loaded when they are required, not when they are added to &packpath / &runtimepath

This is exactly what I'm trying to do with my config. I add as lazy load in my [lazy.nvim](https://github.com/folke/lazy.nvim) and just load it after I call `tree_interface.toggle()` (where it gets required)

>Do this at the very start of your init.lua. By the time you hit your function it will be too late and netrw will have loaded. It's not a regular plugin and starts uncontrollably eagerly.

Thank you for give me this tip.

>I'm also having this problem after migrate to [lazy.nvim](https://github.com/folke/lazy.nvim) package manager.

What I meant was that after I set nvim-tree as lazy load with my new packager manager, the toggle doesn't work and just open the tree when I run it

**Reply by alex-courtis** [reply]
>  What I meant was that after I set nvim-tree as lazy load with my new packager manager, the toggle doesn't work and just open the tree when I run it

There is not anything more I can say. Call setup ONCE and ONCE only, not every time you wish to toggle the tree.

If you can replicate the issue with a [clean room configuration](https://github.com/nvim-tree/nvim-tree.lua/wiki/Troubleshooting#clean-room-replication) please raise a bug with that configuration.

**Comment by bernardassan** [comment, marked as answer]
@nitaicharan To fix this issue, check the recommended `nvim_tree_events_startup` help section. An example of how I used it when I was using paq.nvim can be found [here](https://github.com/Ultra-Code/awesome-neovim/blob/cfd00f9c4d52866b0a5c4578485a0ff33a28ef5d/lua/lazy/tree.lua#L302).

But when using lazy.nvim you shoudn't call `setup fn` outside of the `config fn` of the plugin. check out an example of how to configure [tree](https://github.com/Ultra-Code/awesome-neovim/blob/master/lua/plugins/tree.lua) with lazy.nvim

**Reply by nitaicharan** [reply]
Thank you for bringing it on. I will try it
